channels/api.py

```diff
@@ -476,32 +489,93 @@ def list_comments(self, post_id):
+        comments = more_comments.comments()
+        if isinstance(comments, CommentForest):
+            comments = comments.list()
+
+        # if the number of comments is less than the number of children, it means that the morecomments
+        # object did not return all the comments, so we need to manually add another morecomments
+        # object with the remaining children; not sure why praw does not do it automatically
+        # anyway this seems to happen only with objects that do NOT look like this one:
+        # <MoreComments count=0, children=[]>
+        if len(comments) < len(children):
+            remaining_morecomments = self.init_more_comments(
+                parent_id=parent_id,
+                post_id=post_id,
+                children=children[len(comments):],
+            )
+            comments.append(remaining_morecomments)
+        return comments
+
+    def init_more_comments(self, parent_id, post_id, children):
         """
         Initializes a MoreComments instance from the passed data and fetches channel
 
         Args:
-            comment_fullname(str): the fullname for the comment
-            parent_fullname(str): the fullname of the post
-            count(int): the count of comments
-            children(list(str)): the list of more comments (leave empty continue page links)
+            parent_id (str): the fullname for the comment
+            post_id (str): the id of the post
+            children(list(str)):
+                a list of comment ids
 
         Returns:
             praw.models.MoreComments: the set of more comments
         """
-        submission_id = parent_fullname.split('_', 1)[1]
-        comment_id = comment_fullname.split('_', 1)[1]
-        data = {
-            'id': comment_id,
-            'name': comment_fullname,
-            'parent_id': parent_fullname,
-            'children': children or [],
-            'count': count,
-        }
-        more_comments = more.MoreComments(self.reddit, data)
-        more_comments.submission = self.reddit.submission(submission_id)
+        if parent_id is not None:
+            qualified_parent_id = '{kind}_{parent_id}'.format(
+                kind=get_kind_mapping()['comment'],
+                parent_id=parent_id,
+            )
+        else:
+            qualified_parent_id = '{kind}_{parent_id}'.format(
+                kind=get_kind_mapping()['submission'],
+                parent_id=post_id,
+            )
+
+        more_comments = more.MoreComments(self.reddit, {
+            'children': children,
+            'count': len(children),
+            'parent_id': qualified_parent_id
+        })
+        more_comments.submission = self.reddit.submission(post_id)
+        more_comments.submission.comment_limit = settings.OPEN_DISCUSSIONS_REDDIT_COMMENTS_LIMIT
+
+        original_load_comment = more_comments._load_comment  # pylint: disable=protected-access
+
+        def replacement_load_comment(*args, **kwargs):
```

Comment:
I'd suggest refactoring this into a separate method that returns a replacement method like:

```python
def replace_load_comment(load_comment):
    def replacement_load_comment(*args, **kwargs):
         # impl
    return replacement_load_comment
```

And then the replacement can become a one-liner and we don't have a inlined function anymore:

```python
more_comments._load_comment = replace_load_comment(more_comments._load_comment)
```